Project option: REPLACE_IGNITION_INCLUDE_PATH

scpeters opened the pull request:
# 🎉 New feature

Helps with https://github.com/ignitionrobotics/sdformat/issues/181

## Summary

Currently all cmake projects that use `IgnConfigureProject` are required to use include paths that start with `ignition/${IGN_DESIGNATION}` (i.e. `ignition/math`, `ignition/transport`). We would like to use `ign-cmake2` with `libsdformat`, but its cmake project name and include paths don't follow this pattern. This pull request adds an option to `IgnConfigureProject` to allow more flexibility:

* `INCLUDE_DIR`: option to specify the project include paths, as an alternative to `ignition/${IGN_DESIGNATION}`

## Test it

Look at the `examples/no_ignition_prefix` folder and build with `cmake .. -DBUILDSYSTEM_TESTING=ON && make -j`

## Checklist
- [X] Signed all commits for DCO
- [X] Added tests
- [X] Added example and/or tutorial
- [X] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [X] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [X] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

koonpeng commented:
I'm trying to migrate sdformat to ign-cmake, I found that there are a few more things which are not compatible

* headers are still installed to `ignition/` prefix, it also enforces the `${IGN_DESIGNATION}${PROJECT_VERSION_MAJOR}` pattern, while sdformat uses a slightly different pattern `sdformat-${SDF_VERSION}` (there is a hypen). https://github.com/scpeters/ign-cmake/blob/787cebd91c0ba16562ea6d545b86d5a87d90775c/cmake/IgnPackaging.cmake#L106.
* The master header is named `${IGN_DESIGNATION}.hh`, where `IGN_DESIGNATION` is derived from the project name, which in sdformat's case is `sdformat`, but the master header of sdformat is `sdf.hh`. https://github.com/scpeters/ign-cmake/blob/787cebd91c0ba16562ea6d545b86d5a87d90775c/cmake/IgnUtils.cmake#L724.

I haven't gotten around to migrating tests and docs yet, I will report more issues as I find them.

scpeters commented:
> I'm trying to migrate sdformat to ign-cmake, I found that there are a few more things which are not compatible
> 
> 1. headers are still installed to `ignition/` prefix, it also enforces the `${IGN_DESIGNATION}${PROJECT_VERSION_MAJOR}` pattern, while sdformat uses a slightly different pattern `sdformat-${SDF_VERSION}` (there is a hypen). https://github.com/scpeters/ign-cmake/blob/787cebd91c0ba16562ea6d545b86d5a87d90775c/cmake/IgnPackaging.cmake#L106.
> 2. The master header is named `${IGN_DESIGNATION}.hh`, where `IGN_DESIGNATION` is derived from the project name, which in sdformat's case is `sdformat`, but the master header of sdformat is `sdf.hh`. https://github.com/scpeters/ign-cmake/blob/787cebd91c0ba16562ea6d545b86d5a87d90775c/cmake/IgnUtils.cmake#L724.
> 3. Headers are installed to `${IGN_INCLUDE_INSTALL_DIR_FULL}/${PROJECT_INCLUDE_DIR}`, and the master header is installed one directory up. But sdformat expects the master header to be in the same directory as the other headers.
> 4. The config header must be named `config.hh`, but sdf's config header is named `sdf_config.h`.
> 
> I haven't gotten around to migrating tests and docs yet, I will report more issues as I find them.

I'd like to see if reviewers are ok with this PR as is in its partially useful state or if they want to try to address all of this at. once

cc @adlarkin

adlarkin commented:
Thanks @scpeters and @koonpeng for the things you've discovered so far! Regarding how to proceed with addressing the issues mentioned above, I think we can take the following approach:
1. One PR that gives users options to specify the include directories that are searched for and installed
2. One PR that provides flexibility in specifying the cmake prefix project name
3. One PR that addresses the master header issues (name of the master header and where it's installed)
4. One PR that allows configuration of the name of the config header

Having separate, smaller PRs that focus on solving one issue in particular will make the review process easier, and also give us a better way to track changes being made. If there are no objections to the approach I've suggested, then this PR will need to be broken into 2 PRs because right now, this PR solves part of point 1 (searched include directories) and point 2 above. If help is needed in creating more PRs that solve the remaining issues found by @koonpeng, let me know, and I can help open a few.

scpeters commented:
>     2\. One PR that provides flexibility in specifying the cmake prefix project name

I've split that feature out from this PR into https://github.com/ignitionrobotics/ign-cmake/pull/191

I can rebase this branch after that is merged, if that is not too presumptuous

adlarkin commented:
> I can rebase this branch after that is merged, if that is not too presumptuous

I'm okay with you waiting for #191 and then rebasing. Would you mind updating the PR of this title since you put the `NO_IGNITION_PREFIX` portion in #191?

scpeters commented:
> > I can rebase this branch after that is merged, if that is not too presumptuous
> 
> I'm okay with you waiting for #191 and then rebasing. Would you mind updating the PR of this title since you put the `NO_IGNITION_PREFIX` portion in #191?

done

scpeters commented:
> > > I can rebase this branch after that is merged, if that is not too presumptuous
> > 
> > 
> > I'm okay with you waiting for #191 and then rebasing. Would you mind updating the PR of this title since you put the `NO_IGNITION_PREFIX` portion in #191?
> 
> done

just rebased

adlarkin commented on examples/no_ignition_prefix/CMakeLists.txt:
```diff
@@ -1,7 +1,9 @@
 project(no_ignition_prefix VERSION 0.1.0)
 find_package(ignition-cmake2 REQUIRED)
-ign_configure_project(NO_IGNITION_PREFIX)
+ign_configure_project(
+  INCLUDE_DIR no_ign
+  NO_IGNITION_PREFIX)
```
Can you also update the example's `README` to point out the usage of `INCLUDE_DIR` and what it does?

scpeters replied:
yes, I will do it now

scpeters replied:
https://github.com/ignitionrobotics/ign-cmake/pull/190/commits/3edfd842aa95fe31794995a472a058c654877ceb

scpeters commented:
> Thanks @scpeters and @koonpeng for the things you've discovered so far! Regarding how to proceed with addressing the issues mentioned above, I think we can take the following approach:
> 
>     1. One PR that gives users options to specify the include directories that are searched for and installed
> 
>     2. One PR that provides flexibility in specifying the cmake prefix project name
> 
>     3. One PR that addresses the master header issues (name of the master header and where it's installed)
> 
>     4. One PR that allows configuration of the name of the config header
> 
> 
> Having separate, smaller PRs that focus on solving one issue in particular will make the review process easier, and also give us a better way to track changes being made. If there are no objections to the approach I've suggested, then this PR will need to be broken into 2 PRs because right now, this PR solves part of point 1 (searched include directories) and point 2 above. If help is needed in creating more PRs that solve the remaining issues found by @koonpeng, let me know, and I can help open a few.

I've added a checklist in https://github.com/ignitionrobotics/sdformat/issues/181#issuecomment-962116041 with an additional entry for the prefix to the macros in Export.hh

scpeters commented on cmake/IgnUtils.cmake:
```diff
@@ -936,7 +936,7 @@ function(ign_create_core_library)
   _ign_add_library_or_component(
     LIB_NAME ${PROJECT_LIBRARY_TARGET_NAME}
-    INCLUDE_DIR "ignition/${IGN_DESIGNATION_LOWER}"
+    INCLUDE_DIR "${PROJECT_INCLUDE_DIR}"
     EXPORT_BASE IGNITION_${IGN_DESIGNATION_UPPER}
```
> I noticed that `IGN_DESIGNATION_LOWER` is being replaced here, instead of `IGN_DESIGNATION`. It looks like `IGN_DESIGNATION_LOWER` is the lowercase version of `IGN_DESIGNATION`:
> 
> https://github.com/ignitionrobotics/ign-cmake/blob/4f40889950da9ab57391d8fa5465461a2837f006/cmake/IgnConfigureProject.cmake#L76
> 
> Should we create and use something like `PROJECT_INCLUDE_DIR_LOWER` here for consistency?

`IGN_DESIGNATION` is a special variable that we use in lots of different places with different cases, which is why the `IGN_DESIGNATION_LOWER` variable was created. I don't think we need different case versions of `PROJECT_INCLUDE_DIR` because we just use it in one place

> Also, on a side note, I noticed that `IGN_DESIGNATION_UPPER` is being used in the line below the changes here. While this isn't relate to the include directories work of this PR, I am guessing this is something we'll have to update in another PR if we want to increase flexibility for things like sdformat.

this is the `EXPORT_BASE` issue, which is mentioned in the 3rd checkbox in https://github.com/ignitionrobotics/sdformat/issues/181#issuecomment-962116041. I think we can address that independently

adlarkin commented on examples/no_ignition_prefix/README.md:
```diff
@@ -1,18 +1,34 @@
+
+## Configuring project name
 
 This package uses the `NO_IGNITION_PREFIX` option to `ign_configure_project`
```
```suggestion
This package uses the `NO_IGNITION_PREFIX` option in `ign_configure_project`
```

scpeters replied:
https://github.com/ignitionrobotics/ign-cmake/pull/190/commits/946082f02991e047a688e2b43bfe6123a47f847d

adlarkin commented on examples/no_ignition_prefix/README.md:
```diff
@@ -1,18 +1,34 @@
+
+## Configuring include directory names
+
+This package uses the `INCLUDE_DIR` option to `ign_configure_project`
```
```suggestion
This package uses the `INCLUDE_DIR` option in `ign_configure_project`
```

scpeters replied:
https://github.com/ignitionrobotics/ign-cmake/pull/190/commits/946082f02991e047a688e2b43bfe6123a47f847d

adlarkin reviewed (approved):
LGTM! Just a few nits about grammar before merging

koonpeng commented:
```
diff --git a/cmake/ign_auto_headers.hh.in b/cmake/ign_auto_headers.hh.in
index 42fc964..a078c63 100644
--- a/cmake/ign_auto_headers.hh.in
+++ b/cmake/ign_auto_headers.hh.in
@@ -19,5 +19,5 @@
 // This file is automatically generated by CMake. Changes should instead be
 // made to cmake/ign_auto_headers.hh.in in ignition-cmake
 
-#include <ignition/@IGN_DESIGNATION@/config.hh>
+#include <@PROJECT_INCLUDE_DIR@/config.hh>
 ${ign_headers}
```

Fixes `config.hh` still referencing `ignition/` prefix.

j-rivero commented:
Sorry to jump in a bit late.

>     * `INCLUDE_DIR`: option to specify the project include paths, as an alternative to `ignition/${IGN_DESIGNATION}`

`INCLUDE_DIR` in CMake drives my mind immediately to `INCLUDE_DIRECTORIES` even if we have it under the context of `ign_configure_project`. Maybe we can be a bit more verbose to help code reading with something like: `CUSTOM_INCLUDE_DIR_PATH` or `REPLACE_IGNITION_INCLUDE_PATH` or  other more explicit name about what we are doing with the parameter.

adlarkin commented:
@scpeters would you mind addressing the comments from @j-rivero and @koonpeng above? For the `config.hh` fix, it would be nice to update the example to show that this file can be found without the `ignition` prefix.

scpeters commented:
> Fixes config.hh still referencing ignition/ prefix.

nice catch! fixed in https://github.com/ignitionrobotics/ign-cmake/pull/190/commits/7c12783f43fc414a2d90a6ea193f8769b553b677

scpeters commented:
> Sorry to jump in a bit late.
> 
> > ```
> > * `INCLUDE_DIR`: option to specify the project include paths, as an alternative to `ignition/${IGN_DESIGNATION}`
> > ```
> 
> `INCLUDE_DIR` in CMake drives my mind immediately to `INCLUDE_DIRECTORIES` even if we have it under the context of `ign_configure_project`. Maybe we can be a bit more verbose to help code reading with something like: `CUSTOM_INCLUDE_DIR_PATH` or `REPLACE_IGNITION_INCLUDE_PATH` or other more explicit name about what we are doing with the parameter.

I'll go with `REPLACE_IGNITION_INCLUDE_PATH` as the `ign_configure_project` parameter, but I think I'll keep `PROJECT_INCLUDE_DIR` as the internal variable name

scpeters commented:
> > Sorry to jump in a bit late.
> > > ```
> > > * `INCLUDE_DIR`: option to specify the project include paths, as an alternative to `ignition/${IGN_DESIGNATION}`
> > > ```
> > 
> > 
> > `INCLUDE_DIR` in CMake drives my mind immediately to `INCLUDE_DIRECTORIES` even if we have it under the context of `ign_configure_project`. Maybe we can be a bit more verbose to help code reading with something like: `CUSTOM_INCLUDE_DIR_PATH` or `REPLACE_IGNITION_INCLUDE_PATH` or other more explicit name about what we are doing with the parameter.
> 
> I'll go with `REPLACE_IGNITION_INCLUDE_PATH` as the `ign_configure_project` parameter, but I think I'll keep `PROJECT_INCLUDE_DIR` as the internal variable name

changed in https://github.com/ignitionrobotics/ign-cmake/pull/190/commits/6c2ffe57a0573400532dd45c9da01bdec0f96b2a. It's easy to change the name now, so please speak up now if you have a better idea; it won't hurt my feelings

j-rivero commented:
> changed in [6c2ffe5](https://github.com/ignitionrobotics/ign-cmake/commit/6c2ffe57a0573400532dd45c9da01bdec0f96b2a). It's easy to change the name now, so please speak up now if you have a better idea; it won't hurt my feelings

The change looks good to me, thanks! Since the internal variable was there before I think is fine to leave as it is.

j-rivero reviewed (approved):
ready to go. It would be great if we can write a Changelog entry in this PR but can done at releasing time.

scpeters commented:
> @scpeters would you mind addressing the comments from @j-rivero and @koonpeng above? For the `config.hh` fix, it would be nice to update the example to show that this file can be found without the `ignition` prefix.

https://github.com/ignitionrobotics/ign-cmake/pull/190/commits/6787937391ec245733ea0fc35f13e8f3e3c09af7: I've added the `ign_install_all_headers` to the example and include the generated headers in `AlmostEmpty.cc` to confirm their location

adlarkin reviewed (approved):
Newest changes LGTM, thanks for iterating @scpeters!

> It would be great if we can write a Changelog entry in this PR but can done at releasing time.

+1

osrf-triage commented:
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-01-10/1228/1
